components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/logging/APILogMessageHandler.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.rest.AbstractHandler;
+import org.apache.synapse.rest.RESTConstants;
+
+import java.util.Date;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * A Simple extension handler for message logging for the apis deployed in the API Gateway.
+ * This handler first looks for a sequence named WSO2AM--Ext--[Dir], where [Dir] could be
+ * either In or Out depending on the direction of the message. If such a sequence is found,
+ * It log the information of the particular API. If such an API specific sequence is found, that
+ * is also invoked. If no extension is found either at the global level or at the per API level
+ * this mediator simply returns true.
+ */
+public class APILogMessageHandler extends AbstractHandler {
+
+    private static final String DIRECTION_IN = "In";
+    private static final String DIRECTION_OUT = "Out";
+
+    private static final Log log = LogFactory.getLog(APILogMessageHandler.class);
+
+    private boolean mediate(MessageContext messageContext, String direction) {
+
+        String applicationName = (String) messageContext.getProperty("APPLICATION_NAME");
+        String endUserName = (String) messageContext.getProperty("END_USER_NAME");
+        boolean isLoginRequest = false;
+        String logMessage = "";
+
+        org.apache.axis2.context.MessageContext axisMC =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+
+        if (applicationName != null) {
+            logMessage = " appName=" + applicationName;
+        }
+        if (endUserName != null) {
+            logMessage = logMessage + " , userName=" + endUserName;
+        }
+        Map headers = (Map) axisMC.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        String logID = (String) headers.get("activityID");
+
+        if (DIRECTION_OUT.equals(direction) && logID == null) {
+            try {
+                org.apache.axis2.context.MessageContext inMessageContext =
+                        axisMC.getOperationContext().getMessageContext(WSDL2Constants.MESSAGE_LABEL_IN);
+                if (inMessageContext != null) {
+                    Object inTransportHeaders =
+                            inMessageContext.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+                    if (inTransportHeaders != null) {
+                        String inID = (String) ((Map) inTransportHeaders).get("activityID");
+                        if (inID != null) {
+                            logID = inID;
+                        }
+                    }
+                }
+            } catch (AxisFault axisFault) {
+                log.error("Cannot get Transport headers from Gateway");
```

Comment:
Comment explaining why we're logging and continuing instead of throwing the error.